conjure/enum.go

```diff
@@ -54,32 +56,68 @@ func astForEnum(enumDefinition spec.EnumDefinition, info types.PkgInfo) []astgen
 			Values:  []astgen.ASTExpr{expression.StringVal(currVal.Value)},
 		})
 	}
-	vals = append(vals, &astspec.Value{
-		Names:  []string{enumName + toCamelCase(unknownEnumValue)},
-		Type:   expression.Type(enumName),
-		Values: []astgen.ASTExpr{expression.StringVal(unknownEnumValue)},
-	})
 	valsDecl := &decl.Const{Values: vals}
 
 	unmarshalDecl := enumUnmarshalTextAST(enumDefinition, info)
 
 	return []astgen.ASTDecl{typeDef, valsDecl, unmarshalDecl}
 }
 
+func astForEnumPattern(info types.PkgInfo) astgen.ASTDecl {
+	info.AddImports("regexp")
+	matchString := expression.NewCallFunction("regexp", "MustCompile", expression.StringVal(enumValuePattern))
+	return &decl.Var{
+		Name:  enumPatternVarName,
+		Value: matchString,
+	}
+}
+
 func enumUnmarshalTextAST(e spec.EnumDefinition, info types.PkgInfo) astgen.ASTDecl {
+	mapStringInterface := expression.Type(types.NewMapType(types.String, types.Any).GoType(info))
 	toCamelCase := varcaser.Caser{From: varcaser.ScreamingSnakeCase, To: varcaser.UpperCamelCase}.String
 
 	info.AddImports("strings")
+	info.AddImports("github.com/palantir/conjure-go-runtime/v2/conjure-go-contract/errors")
+	info.AddImports("github.com/palantir/witchcraft-go-error")
+	info.AddImports("github.com/palantir/witchcraft-go-params")
+
 	switchStmt := &statement.Switch{
-		Expression: expression.NewCallFunction("strings", "ToUpper", expression.NewCallExpression(expression.StringType, expression.VariableVal(dataVarName))),
+		Init:       statement.NewAssignment(expression.VariableVal(enumUpperVarName), token.DEFINE, expression.NewCallFunction("strings", "ToUpper", expression.NewCallExpression(expression.StringType, expression.VariableVal(dataVarName)))),
+		Expression: expression.VariableVal(enumUpperVarName),
 		Cases: []statement.CaseClause{
 			// default case
 			{
-				Body: []astgen.ASTStmt{statement.NewAssignment(
-					expression.NewUnary(token.MUL, expression.VariableVal(enumReceiverName)),
-					token.ASSIGN,
-					expression.VariableVal(e.TypeName.Name+toCamelCase(unknownEnumValue)),
-				),
+				Body: []astgen.ASTStmt{
+					&statement.If{
+						Cond: expression.NewUnary(token.NOT,
+							expression.NewCallExpression(
+								expression.NewSelector(expression.VariableVal(enumPatternVarName), "MatchString"),
+								expression.VariableVal(enumUpperVarName),
+							),
+						),
+						Body: []astgen.ASTStmt{
+							statement.NewReturn(
+								expression.NewCallFunction("werror", "Convert",
```

Comment:
What is the value in adding in the additional `werror.Convert` given the `NewInvalidArgument` will return a proper error?

Reply:
Adam asked the same question on Slack. This might be exposing a deficiency with how we implemented the conjure error, but because it's not a werror it doesn't have a stacktrace. I added the werror.Convert as low as possible to preserve the stack frame(s) between here and the next werror.Wrap.